core/shared/src/main/scala/spire/algebra/UniqueFactorizationDomain.scala

```diff
@@ -0,0 +1,47 @@
+package spire
+package algebra
+
+import spire.math.{Integral, SafeLong}
+
+/** A unique factorization domain is a commutative ring in which each element can be written
+  * as a product of prime elements and a unit.
+  * 
+  * This trait is outside the commutative ring hierarchy, because the factorization algorithms
+  * are costly. Another reason: in some cases, a deliberate choice should be made by the user,
+  * for example to use probabilistic algorithms with a specified probability of failure.
+  */
+trait UniqueFactorizationDomain[@sp(Byte, Short, Int, Long) A] extends Any {
+  /** Tests whether the given nonzero element is prime. */
+  def isPrime(a: A): Boolean
+
+  /** Returns the factors of the given nonzero element. */
+  def factor(a: A): UniqueFactorizationDomain.Decomposition[A]
+}
+
+object UniqueFactorizationDomain {
+
+  trait Decomposition[@sp(Byte, Short, Int, Long) A] {
+    def unit: A
+    def elements: Map[A, Int]
+  }
+
+  def apply[A](implicit ev: UniqueFactorizationDomain[A]): UniqueFactorizationDomain[A] = ev
+
+  case class WrapDecomposition[A:CRing](safeLongFactors: spire.math.prime.Factors) extends Decomposition[A] {
```

Comment:
Instead of a `WrapDecomposition`, maybe we just have a `map[B](f: A => B): Decomposition[B]` method in `Decomposition`? Then we can just define the `Integral` case as `def factor(a: A): Decomposition[A] = SafeLong(A.toBigInt(a)).factor.map(CRing[A].fromBigInt(_))`.